cherrypick-2.0: fixes for disabling new-version pings

dt opened the pull request:
went back and forth on cherrypicking the first commit removing the pointless setting: initially I was reluctant, as it is unlikely we'd include something like it in a hypothetical 2.0.1 patch release, but in conjunction with the other fixes here, it helps simplify the ux for diagnostics / phone-home behavior.

dt committed:
server: don't crash on empty update check or diagnostic report overrides

You can override the updates check or diagnoatics report endpoints via env vars, but overriding them to empty would result in a nil pointer deference crash when the server attempted to check for new versions or send a diagnostics report.
Fixes #22967.

Release note (bug fix): do not crash if COCKROACH_UPDATE_CHECK_URL or COCKROACH_USAGE_REPORT_URL overrides are empty.

dt committed:
server: skip check for new versions if diagnostics reports are disabled

If a user has disabled diagnostics reporting, we should assume that that expressed a preference that we not “phone-home” at all, including for the new-version checks.

Release note (general change): disabling diagnostics reporting also disables new version notification checks.

dt committed:
server: remove diagnostics.reporting.report_metrics setting

This setting is duplicative with diagnostics.reporting.enabled — the only reason to set it would be to disable the diagnostics reports while leaving other diagnostics enabled — i.e. crash reporting and new-version checks, but it’s easier to explain if there’s just a single flag to toggle.

Release note (general change): remove the cluster setting diagnostics.reporting.report_metrics.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/23007)
<!-- Reviewable:end -->

dt committed:
settings: introduce retired setting state

Addresses #21129: eagerly introducing a migration to delete a remove
setting's value is risky as older nodes are likely still using that
value. We should delay any removal until after the cluster upgrade
process is completed. However we can avoid accidental re-use of the
name, and the log spam of an unknown value, by maintaining a list of
retired settings. In the future, we can add migrations after major
versions that flush this list in single, version-protected migration,
cleaning up old values.

Release note: none.

dt committed:
migrations: back out settings purges

per discussion in #21129 and #23130, we want to avoid removing settings too soon as they may still be in use.

Release note: none.

dt commented:
thanks, added the retired settings fixes to this, ready for another look!

knz commented:
:+1: :shipit:

---

Reviewed 2 of 2 files at r1, 2 of 2 files at r2, 1 of 1 files at r3, 3 of 3 files at r4, 3 of 3 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23007#-:-L6NUAq8MgFptjRlSA_8:bef72r4)*
<!-- Sent from Reviewable.io -->
